### ISSUE TITLE
Viewport out-of-range when first constructing certain UIs rendered with egui-wgpu callbacks

### ISSUE THREAD

**Issue by msparkles** [emilk/egui#3759]
**Describe the bug**

Viewport out-of-range, causing a validation error, when first constructing certain UIs rendered with egui-wgpu callbacks.
This only happens on the first frame, the fix would simply be skipping the callback and set viewport command when the viewport rect is outside of the display.

**To Reproduce**
1. Make a Bottom panel with a scroll area inside
2. Add egui-wgpu Callbacks to it

**Expected behavior**
The item to be skipped if the viewport is outside of display.

**Additional context**
https://github.com/automancy/automancy/blob/main/src/bin/automancy/gui/tile_selection.rs

(Remove the following from the code to actually get the panic:
```rs
        if !ui.ctx().screen_rect().contains_rect(rect) {
            continue;
        }
```
)

**Comment by msparkles** [comment]
This needs an egui fix because the scissor could be smaller than the viewport.

**Comment by emilk** [comment]
Is it the `PaintCallback::rect` that is wrong, or the clip rect, or both? What are their values compared to the viewport rect?

Why does it only happen on the first frame?

> the fix would simply be skipping the callback and set viewport command when the viewport rect is outside of the display.

what about when it is partially outside the frame?

**Comment by msparkles** [comment]
Actually, we just thought about it for a bit- the problem should be that the scissor isn't being correctly set..

**Comment by msparkles** [comment]
> Why does it only happen on the first frame?
Regarding this, we can only assume it's some placement negotiation oddities that's happening, and that's beyond our knowledge to really understand..

**Comment by msparkles** [comment]
Here's a stacktrace:

~~Don't ask about the path, it's an in-joke...~~

```
The application panicked (crashed).
Message:  wgpu error: Validation Error

Caused by:
    In a RenderPass
      note: encoder = `Render Encoder`
    In a set_viewport command
    Viewport has invalid rect Rect { x: 20.0, y: 600.0, w: 80.0, h: 0.0 }; origin and/or size is less than or equal to 0, and/or is not contained in the render target Extent3d { width: 800, height: 600, depth_or_array_layers: 1 }


Location: /mnt/cum/.cargo/registry/src/index.crates.io-6f17d22bba15001f/wgpu-0.18.0/src/backend/direct.rs:3111

  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ BACKTRACE ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
   1: color_eyre::config::PanicHook::panic_report::h26173db51b99cbec
      at /mnt/cum/.cargo/registry/src/index.crates.io-6f17d22bba15001f/color-eyre-0.6.2/src/config.rs:1002
   2: automancy::main::{{closure}}::h36f4ef77593b11c3
      at /mnt/cum/automancy/src/bin/automancy/main.rs:98
   3: <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call::h1f8f335eaa9cfaee
      at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/alloc/src/boxed.rs:2021
   4: std::panicking::rust_panic_with_hook::h2b5517d590cab22e
      at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/panicking.rs:783
   5: std::panicking::begin_panic_handler::{{closure}}::h233112c06e0ef43e
      at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/panicking.rs:657
   6: std::sys_common::backtrace::__rust_end_short_backtrace::h6e893f24d7ebbff8
      at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/sys_common/backtrace.rs:170
   7: rust_begin_unwind<unknown>
      at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/panicking.rs:645
   8: core::panicking::panic_fmt::hbf0e066aabfa482c
      at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/panicking.rs:72
   9: wgpu::backend::direct::default_error_handler::h5ef848bf68441ded
      at /mnt/cum/.cargo/registry/src/index.crates.io-6f17d22bba15001f/wgpu-0.18.0/src/backend/direct.rs:3111
  10: core::ops::function::Fn::call::h0291ab9979cc497f
      at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/ops/function.rs:79
  11: <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call::h804c391efe5c1bc2
      at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/alloc/src/boxed.rs:2021
  12: wgpu::backend::direct::ErrorSinkRaw::handle_error::hcaa1097aa20eeebf
      at /mnt/cum/.cargo/registry/src/index.crates.io-6f17d22bba15001f/wgpu-0.18.0/src/backend/direct.rs:3097
  13: wgpu::backend::direct::Context::handle_error::h7897817e248461e0
      at /mnt/cum/.cargo/registry/src/index.crates.io-6f17d22bba15001f/wgpu-0.18.0/src/backend/direct.rs:333
  14: <wgpu::backend::direct::Context as wgpu::context::Context>::command_encoder_end_render_pass::h572aa01466224a17
      at /mnt/cum/.cargo/registry/src/index.crates.io-6f17d22bba15001f/wgpu-0.18.0/src/backend/direct.rs:1984
  15: <T as wgpu::context::DynContext>::command_encoder_end_render_pass::h7f0744be1e9a14c0
      at /mnt/cum/.cargo/registry/src/index.crates.io-6f17d22bba15001f/wgpu-0.18.0/src/context.rs:2830
  16: <wgpu::RenderPass as core::ops::drop::Drop>::drop::h3e69d6eeb8e8e0bd
      at /mnt/cum/.cargo/registry/src/index.crates.io-6f17d22bba15001f/wgpu-0.18.0/src/lib.rs:4219
  17: core::ptr::drop_in_place<wgpu::RenderPass>::he38f93a94441ca19
      at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/ptr/mod.rs:498
  18: automancy::renderer::Renderer::inner_render::h4c4dbe4368e1f6c4
      at /mnt/cum/automancy/src/bin/automancy/renderer.rs:662
  19: automancy::renderer::Renderer::render::h0a05bfb3d74231cc
      at /mnt/cum/automancy/src/bin/automancy/renderer.rs:325
  20: automancy::event::render::hec5ca09b0883714a
      at /mnt/cum/automancy/src/bin/automancy/event.rs:307
  21: automancy::event::on_event::h8c1f8892e3200a19
      at /mnt/cum/automancy/src/bin/automancy/event.rs:366
  22: automancy::main::{{closure}}::h8ab05b7859ee6fae
      at /mnt/cum/automancy/src/bin/automancy/main.rs:216
  23: core::ops::function::impls::<impl core::ops::function::FnMut<A> for &mut F>::call_mut::h131e1661d2b76dff
      at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/ops/function.rs:294
  24: core::ops::function::impls::<impl core::ops::function::FnMut<A> for &mut F>::call_mut::he84c07f7035798f6
      at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/ops/function.rs:294
  25: winit::platform_impl::platform::x11::EventLoop<T>::single_iteration::h2a0b1c8ce49969a6
      at /mnt/cum/.cargo/registry/src/index.crates.io-6f17d22bba15001f/winit-0.29.8/src/platform_impl/linux/x11/mod.rs:608
  26: winit::platform_impl::platform::x11::EventLoop<T>::poll_events_with_timeout::hcc07a7bb076caca1
      at /mnt/cum/.cargo/registry/src/index.crates.io-6f17d22bba15001f/winit-0.29.8/src/platform_impl/linux/x11/mod.rs:547
  27: winit::platform_impl::platform::x11::EventLoop<T>::pump_events::haceecba728870926
      at /mnt/cum/.cargo/registry/src/index.crates.io-6f17d22bba15001f/winit-0.29.8/src/platform_impl/linux/x11/mod.rs:453
  28: winit::platform_impl::platform::x11::EventLoop<T>::run_on_demand::h0522eea49eee57fa
      at /mnt/cum/.cargo/registry/src/index.crates.io-6f17d22bba15001f/winit-0.29.8/src/platform_impl/linux/x11/mod.rs:414
  29: winit::platform_impl::platform::EventLoop<T>::run_on_demand::h463c92fb95c64045
      at /mnt/cum/.cargo/registry/src/index.crates.io-6f17d22bba15001f/winit-0.29.8/src/platform_impl/linux/mod.rs:821
  30: winit::platform_impl::platform::EventLoop<T>::run::he2d5eec979356426
      at /mnt/cum/.cargo/registry/src/index.crates.io-6f17d22bba15001f/winit-0.29.8/src/platform_impl/linux/mod.rs:814
  31: winit::event_loop::EventLoop<T>::run::hf25fb5a8c62709be
      at /mnt/cum/.cargo/registry/src/index.crates.io-6f17d22bba15001f/winit-0.29.8/src/event_loop.rs:249
  32: automancy::main::h1e3686277562dfbd
      at /mnt/cum/automancy/src/bin/automancy/main.rs:211
  33: core::ops::function::FnOnce::call_once::ha951cb4a3447fddb
      at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/ops/function.rs:250
  34: std::sys_common::backtrace::__rust_begin_short_backtrace::hfd524f999dc1925c
      at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/sys_common/backtrace.rs:154
  35: std::rt::lang_start::{{closure}}::h31e6c6fbf77b973b
      at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/rt.rs:167
  36: core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &F>::call_once::hd95060ecd5e1ca24
      at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/ops/function.rs:284
  37: std::panicking::try::do_call::h6e8cf51db32a6e4b
      at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/panicking.rs:552
  38: std::panicking::try::h3a52eefe24fe3c29
      at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/panicking.rs:516
  39: std::panic::catch_unwind::h24c28c23c02c3841
      at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/panic.rs:142
  40: std::rt::lang_start_internal::{{closure}}::h705d3c9cbc06ef47
      at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/rt.rs:148
  41: std::panicking::try::do_call::ha21f52ba13158470
      at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/panicking.rs:552
  42: std::panicking::try::h5581346bf6aeb1f8
      at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/panicking.rs:516
  43: std::panic::catch_unwind::h7919645a6b72e25b
      at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/panic.rs:142
  44: std::rt::lang_start_internal::h12de51168669836e
      at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/rt.rs:148
  45: std::rt::lang_start::h41a817c3dd481e40
      at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/rt.rs:166
  46: __libc_start_main<unknown>
      at <unknown source file>:<unknown line>
  47: _start<unknown>
      at <unknown source file>:<unknown line>
```

**Comment by msparkles** [comment]
Gonna work on this right now and implement a fix. \o/

**Comment by msparkles** [comment]
Okay, to recap everything:

- Apparently there's a discussion? on this-- We came from Vulkan, we used to render outside of the render attachment and clip stuff with scissor, so that we can render half an object-- https://github.com/gpuweb/gpuweb/issues/373 apparently this is an.... issue in wgpu?
- Apparently there's still a minor bug in egui where it would try to set a viewport of width and/or height of 0, just plug this before the set viewport command- this is unrelated to this issue

```rs
if viewport_px.width_px <= 0 || viewport_px.height_px <= 0 {
    continue;
}
```

- Part of this is an user error on our part- with that piece of code patched onto egui, the issue now is this:

![image](https://github.com/emilk/egui/assets/33348966/561007dd-39f8-40b9-b1d7-1b723c1bd62f)

Aka. The Thing Is Not Supposed to be Squished- but this is apparently wgpu's issue now?

**Comment by msparkles** [comment]
CC: @cwfitzgerald You're the wgpu person right? :P 

tldr: apparently WebGPU? says that, you should be able to set viewport to be outside of attachment, but wgpu doesn't allow that

**Comment by emilk** [comment]
I think @Wumpf has thought about this before when we implemented a similar thing in Rerun

**Comment by Wumpf** [comment]
_Today_ WebGPU specification **requires** to fail validation on the device timeline if the viewport is outside of the render targets, see https://www.w3.org/TR/webgpu/#dom-gpurenderpassencoder-setviewport so wgpu behaves correctly here.

With that am I correct that the only thing wrong on the egui side is that it should discard zero area viewports?
Not sure actually whether it should still invoke egui render callbacks 🤔, I guess it should.

**Comment by msparkles** [comment]
> Today, WebGPU specification requires to fail validation on the device timeline if the viewport is outside of the render targets

That is, rather annoying.. is there any other way to achieve this rendering then? Partially out of range and then clipped into the screen.

> With that am I correct that the only thing wrong on the egui side is that it should discard zero area viewports?

Yeah.

Imo it should skip the whole callback, because if not updated it'll just use the older viewport, which is gonna override other stuff.

**Comment by cwfitzgerald** [comment]
> That is, rather annoying.. is there any other way to achieve this rendering then? Partially out of range and then clipped into the screen.

Intersection of your desired rect and the screen viewport rect.

**Comment by emilk** [comment]
Shapes with empty clip rects are already pruned by the `epaint` tessellator before the results are handed to the rendering backend: https://github.com/emilk/egui/blob/34c341444f45686f5ae115884bb6232f15a2e1b5/crates/epaint/src/tessellator.rs#L1140-L1142

**Comment by msparkles** [comment]
In here..

https://github.com/emilk/egui/blob/34c341444f45686f5ae115884bb6232f15a2e1b5/crates/egui-wgpu/src/renderer.rs#L453

The rect is checked during callback...

...but that rect isn't clamped to screen size yet.
